collections/threads.go

```diff
@@ -0,0 +1,204 @@
+package collections
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/textileio/go-threads/core/thread"
+	"github.com/textileio/textile/api/common"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+var (
+	threadNameRx *regexp.Regexp
+
+	ErrInvalidThreadName = fmt.Errorf("name may only contain alphanumeric characters or non-consecutive hyphens, and cannot begin or end with a hyphen")
+)
+
+func init() {
+	threadNameRx = regexp.MustCompile(`^[A-Za-z0-9]+(?:[-][A-Za-z0-9]+)*$`)
+}
+
+type Thread struct {
+	ID    thread.ID
+	Owner crypto.PubKey
+	Name  string
+	Key   string
+}
+
+type Threads struct {
+	col *mongo.Collection
+}
+
+func NewThreads(ctx context.Context, db *mongo.Database) (*Threads, error) {
+	d := &Threads{col: db.Collection("threads")}
+	_, err := d.col.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys: bson.D{{"_id.owner", 1}, {"name", 1}},
+			Options: options.Index().SetUnique(true).
+				SetPartialFilterExpression(bson.D{{"name", bson.M{"$exists": 1}}}).
```

Comment:
Some fancy partial index filtering... only include the compound index if name is not empty.